Package Plugin Integration Tests Flakes

**Bug description**
Package Plugin Integration Tests Flakes

```
[2022-07-14T01:47:35.705Z] $ tanzu package installed delete test-pkg --namespace test-ns --poll-interval 20s --poll-timeout 10m0s -y --kubeconfig /home/kubo/.kube/config 

[2022-07-14T01:47:35.964Z] package 'test-pkg' is not installed in namespace 'test-ns'.

[2022-07-14T01:47:35.964Z] $ tanzu package repository delete carvel-test --namespace test-ns --wait=true --poll-interval 20s --poll-timeout 10m0s -y --kubeconfig /home/kubo/.kube/config 

[2022-07-14T01:47:36.535Z] package repository 'carvel-test' does not exist in namespace 'test-ns'

[2022-07-14T01:47:36.535Z] $ tanzu secret registry delete test-secret --namespace default -y --kubeconfig /home/kubo/.kube/config 

[2022-07-14T01:47:37.102Z] 

[2022-07-14T01:47:37.102Z]  registry secret 'test-secret' does not exist in namespace 'default'

[2022-07-14T01:47:42.373Z] $ tanzu package repository list --namespace test-ns --output json --kubeconfig /home/kubo/.kube/config 

[2022-07-14T01:47:42.373Z] Error: failed to check for the availability of 'packaging.carvel.dev' API: failed to discover unmatched GroupVersionResources: the server is currently unable to handle the request

[2022-07-14T01:47:42.373Z] 

[2022-07-14T01:47:42.373Z] ✖  exit status 1 

[2022-07-14T01:47:42.373Z] Error: exit status 1

[2022-07-14T01:47:42.373Z] <nil>

[2022-07-14T01:47:42.373Z] <nil>

[2022-07-14T01:47:57.312Z] $ tanzu package repository list --namespace test-ns --output json --kubeconfig /home/kubo/.kube/config 

[2022-07-14T01:47:57.312Z] Error: failed to check for the availability of 'packaging.carvel.dev' API: failed to discover unmatched GroupVersionResources: the server is currently unable to handle the request

[2022-07-14T01:47:57.312Z] 

[2022-07-14T01:47:57.312Z] ✖  exit status 1 

[2022-07-14T01:47:57.312Z] Error: exit status 1

[2022-07-14T01:47:57.312Z] <nil>

[2022-07-14T01:47:57.312Z] <nil>

[2022-07-14T01:48:15.398Z] $ tanzu package repository list --namespace test-ns --output json --kubeconfig /home/kubo/.kube/config 

[2022-07-14T01:48:15.398Z] Error: failed to check for the availability of 'packaging.carvel.dev' API: failed to discover unmatched GroupVersionResources: the server is currently unable to handle the request

[2022-07-14T01:48:15.398Z] 

[2022-07-14T01:48:15.398Z] ✖  exit status 1 

[2022-07-14T01:48:15.398Z] Error: exit status 1

[2022-07-14T01:48:15.398Z] <nil>

[2022-07-14T01:48:15.398Z] <nil>

[2022-07-14T01:48:30.365Z] $ tanzu package repository list --namespace test-ns --output json --kubeconfig /home/kubo/.kube/config 

[2022-07-14T01:48:30.365Z] Error: failed to check for the availability of 'packaging.carvel.dev' API: failed to discover unmatched GroupVersionResources: the server is currently unable to handle the request

[2022-07-14T01:48:30.365Z] 

[2022-07-14T01:48:30.365Z] ✖  exit status 1 

[2022-07-14T01:48:30.365Z] Error: exit status 1

[2022-07-14T01:48:30.365Z] <nil>

[2022-07-14T01:48:30.365Z] <nil>

[2022-07-14T01:48:30.365Z] STEP: cleanup previous package installation

[2022-07-14T01:48:30.365Z] STEP: cleanup previous package repository installation

[2022-07-14T01:48:30.365Z] STEP: cleanup previous secret installation

[2022-07-14T01:48:30.365Z] STEP: set up private registry in cluster

[2022-07-14T01:48:30.365Z] configmap/kapp-controller-config patched

[2022-07-14T01:48:30.365Z] namespace/registry created

[2022-07-14T01:48:30.365Z] configmap/registry-contents created

[2022-07-14T01:48:30.365Z] secret/htpasswd-auth created

[2022-07-14T01:48:30.365Z] secret/registry-certs created

[2022-07-14T01:48:30.365Z] service/registry-svc created

[2022-07-14T01:48:30.365Z] deployment.apps/self-signed-registry created

[2022-07-14T01:48:30.365Z] STEP: restart kapp-controller pod to pick up configmap changes

[2022-07-14T01:48:30.365Z] STEP: make sure package API is available after kapp-controller restart

[2022-07-14T01:48:30.365Z] • Failure [54.682 seconds]

[2022-07-14T01:48:30.365Z] Package plugin integration test

[2022-07-14T01:48:30.365Z] /home/kubo/tanzu-framework/pkg/v1/tkg/test/tkgpackageclient/package_plugin_integration_test.go:134

[2022-07-14T01:48:30.365Z]   testing package plugin on workload cluster

[2022-07-14T01:48:30.365Z]   /home/kubo/tanzu-framework/pkg/v1/tkg/test/tkgpackageclient/package_plugin_integration_test.go:368

[2022-07-14T01:48:30.365Z]     should pass all checks on workload cluster [It]

[2022-07-14T01:48:30.365Z]     /home/kubo/tanzu-framework/pkg/v1/tkg/test/tkgpackageclient/package_plugin_integration_test.go:373

[2022-07-14T01:48:30.365Z] 

[2022-07-14T01:48:30.365Z]     Unexpected error:

[2022-07-14T01:48:30.365Z]         <*exec.ExitError | 0xc000eb6060>: {

[2022-07-14T01:48:30.365Z]             ProcessState: {

[2022-07-14T01:48:30.365Z]                 pid: 42336,

[2022-07-14T01:48:30.365Z]                 status: 256,

[2022-07-14T01:48:30.365Z]                 rusage: {

[2022-07-14T01:48:30.365Z]                     Utime: {Sec: 0, Usec: 146362},

[2022-07-14T01:48:30.365Z]                     Stime: {Sec: 0, Usec: 55458},

[2022-07-14T01:48:30.365Z]                     Maxrss: 71308,

[2022-07-14T01:48:30.365Z]                     Ixrss: 0,

[2022-07-14T01:48:30.365Z]                     Idrss: 0,

[2022-07-14T01:48:30.365Z]                     Isrss: 0,

[2022-07-14T01:48:30.365Z]                     Minflt: 10736,

[2022-07-14T01:48:30.365Z]                     Majflt: 3,

[2022-07-14T01:48:30.365Z]                     Nswap: 0,

[2022-07-14T01:48:30.365Z]                     Inblock: 736,

[2022-07-14T01:48:30.365Z]                     Oublock: 8,

[2022-07-14T01:48:30.365Z]                     Msgsnd: 0,

[2022-07-14T01:48:30.365Z]                     Msgrcv: 0,

[2022-07-14T01:48:30.365Z]                     Nsignals: 0,

[2022-07-14T01:48:30.365Z]                     Nvcsw: 793,

[2022-07-14T01:48:30.365Z]                     Nivcsw: 25,

[2022-07-14T01:48:30.365Z]                 },

[2022-07-14T01:48:30.365Z]             },

[2022-07-14T01:48:30.365Z]             Stderr: nil,

[2022-07-14T01:48:30.365Z]         }

[2022-07-14T01:48:30.365Z]         exit status 1

[2022-07-14T01:48:30.365Z]     occurred

[2022-07-14T01:48:30.365Z] 

[2022-07-14T01:48:30.365Z]     /home/kubo/tanzu-framework/pkg/v1/tkg/test/tkgpackageclient/package_plugin_integration_test.go:489

[2022-07-14T01:48:30.365Z] ------------------------------

[2022-07-14T01:48:30.365Z] 

[2022-07-14T01:48:30.365Z] 

[2022-07-14T01:48:30.365Z] Summarizing 2 Failures:

[2022-07-14T01:48:30.365Z] 

[2022-07-14T01:48:30.365Z] [Fail] Package plugin integration test testing package plugin on management cluster [It] should pass all checks on management cluster 

[2022-07-14T01:48:30.365Z] /home/kubo/tanzu-framework/pkg/v1/tkg/test/tkgpackageclient/package_plugin_integration_test.go:548

[2022-07-14T01:48:30.365Z] 

[2022-07-14T01:48:30.365Z] [Fail] Package plugin integration test testing package plugin on workload cluster [It] should pass all checks on workload cluster 

[2022-07-14T01:48:30.365Z] /home/kubo/tanzu-framework/pkg/v1/tkg/test/tkgpackageclient/package_plugin_integration_test.go:489

[2022-07-14T01:48:30.365Z] 

[2022-07-14T01:48:30.365Z] Ran 2 of 2 Specs in 66.347 seconds

[2022-07-14T01:48:30.365Z] FAIL! -- 0 Passed | 2 Failed | 0 Pending | 0 Skipped

[2022-07-14T01:48:30.365Z] --- FAIL: TestTkgpackageclient (66.35s)

[2022-07-14T01:48:30.365Z] FAIL

[2022-07-14T01:48:30.365Z] 

[2022-07-14T01:48:30.365Z] Ginkgo ran 1 suite in 3m59.319596768s

[2022-07-14T01:48:30.365Z] Test Suite Failed

[2022-07-14T01:48:30.365Z] 
```

**Affected product area (please put an X in all that apply)**

- ( ) APIs
- ( ) Addons
- ( ) CLI
- ( ) Docs
- ( ) IAM
- ( ) Installation
- ( ) Plugin
- ( ) Security
- ( ) Test and Release
- ( ) User Experience
- ( ) Developer Experience

**Expected behavior**

**Steps to reproduce the bug**

**Version** (include the SHA if the version is not obvious)

**Environment where the bug was observed (cloud, OS, etc)**

**Relevant Debug Output (Logs, manifests, etc)**

--- a/pkg/v1/tkg/test/tkgpackageclient/config/package_plugin_config.yaml
+++ b/pkg/v1/tkg/test/tkgpackageclient/config/package_plugin_config.yaml
@@ -2,6 +2,7 @@
 use-existing-cluster: false
 mc-cluster-name: ""
 wlc-cluster-name: ""
+wlc-cluster-namespace: default
 mc-kubeconfig-Path: ""
 wlc-kubeconfig-Path: ""
 namespace: "test-ns"

--- a/pkg/v1/tkg/test/tkgpackageclient/package_plugin_integration_test.go
+++ b/pkg/v1/tkg/test/tkgpackageclient/package_plugin_integration_test.go
@@ -55,6 +55,7 @@ type PackagePluginConfig struct {
 	RepositoryPrivateTag      string `json:"repository-private-tag"`
 	ClusterNameMC             string `json:"mc-cluster-name"`
 	ClusterNameWLC            string `json:"wlc-cluster-name"`
+	ClusterNamespaceWLC       string `json:"wlc-cluster-namespace"`
 	KubeConfigPathMC          string `json:"mc-kubeconfig-Path"`
 	KubeConfigPathWLC         string `json:"wlc-kubeconfig-Path"`
 	WithValueFile             bool   `json:"with-value-file"`
@@ -158,6 +159,10 @@ var _ = Describe("Package plugin integration test", func() {
 
 		if config.ClusterNameWLC == "" {
 			config.ClusterNameWLC = fmt.Sprintf(framework.TkgDefaultClusterPrefix + "wlc-test-pkg-plugin")
+		}
+
+		if config.ClusterNamespaceWLC == "" {
+			config.ClusterNamespaceWLC = "default"
 		}
 
 		if config.KubeConfigPathMC == "" {
@@ -370,8 +375,14 @@ var _ = Describe("Package plugin integration test", func() {
 			packagePlugin = packagelib.NewPackagePlugin(config.KubeConfigPathWLC, pollInterval, pollTimeout, "json", "", 0)
 			secretPlugin = secretlib.NewSecretPlugin(config.KubeConfigPathWLC, pollInterval, pollTimeout, "json", 0)
 		})
+		AfterEach(func() {
+			paused := false
+			pauseKappControllerPackage(config.ClusterNameWLC, config.ClusterNamespaceWLC, config.KubeConfigPathMC, config.ClusterNameMC, paused)
+		})
 		It("should pass all checks on workload cluster", func() {
 			cleanup()
+			paused := true
+			pauseKappControllerPackage(config.ClusterNameWLC, config.ClusterNamespaceWLC, config.KubeConfigPathMC, config.ClusterNameMC, paused)
 			setUpPrivateRegistry(config.KubeConfigPathWLC, config.ClusterNameWLC)
 			testHelper()
 			log.Info("Successfully finished package plugin integration tests on workload cluster")
@@ -417,12 +428,26 @@ func setUpPrivateRegistry(kubeconfigPath, clusterName string) {
 	err = command.RunAndRedirectOutput(context.Background())
 	Expect(err).ToNot(HaveOccurred())
 
-	command = exec.NewCommand(
-		exec.WithCommand("kubectl"),
-		exec.WithArgs("apply", "-f", "config/assets/registry-contents.yml", "--context", kubeCtx, "--kubeconfig", kubeconfigPath),
-		exec.WithStdout(GinkgoWriter),
-	)
-	err = command.RunAndRedirectOutput(context.Background())
+	backOff := wait.Backoff{
+		Steps:    10,
+		Duration: 15 * time.Second,
+		Factor:   1.0,
+		Jitter:   0.1,
+	}
+
+	err = retry.OnError(
+		backOff,
+		func(err error) bool {
+			return err != nil
+		},
+		func() error {
+			command = exec.NewCommand(
+				exec.WithCommand("kubectl"),
+				exec.WithArgs("apply", "-f", "config/assets/registry-contents.yml", "--context", kubeCtx, "--kubeconfig", kubeconfigPath),
+				exec.WithStdout(GinkgoWriter),
+			)
+			return command.RunAndRedirectOutput(context.Background())
+		})
 	Expect(err).ToNot(HaveOccurred())
 
 	command = exec.NewCommand(
@@ -467,13 +492,6 @@ func setUpPrivateRegistry(kubeconfigPath, clusterName string) {
 	}
 
 	By("make sure package API is available after kapp-controller restart")
-	backOff := wait.Backoff{
-		Steps:    4,
-		Duration: 15 * time.Second,
-		Factor:   1.0,
-		Jitter:   0.1,
-	}
-
 	err = retry.OnError(
 		backOff,
 		func(err error) bool {
@@ -507,6 +525,19 @@ func setUpPrivateRegistry(kubeconfigPath, clusterName string) {
 			}
 			return nil
 		})
+	Expect(err).ToNot(HaveOccurred())
+}
+
+func pauseKappControllerPackage(clusterName, clusterNamespace, mgmtClusterKubeconfigPath, mgmtClusterName string, pause bool) {
+	By("pausing kapp controller app")
+	mgmtClusterKubeCtx := mgmtClusterName + "-admin@" + mgmtClusterName
+
+	command := exec.NewCommand(
+		exec.WithCommand("kubectl"),
+		exec.WithArgs("patch", "app", fmt.Sprintf("%s-kapp-controller", clusterName), "-n", clusterNamespace, "--type", "merge", "-p", fmt.Sprintf("{\"spec\":{\"paused\":%t}}", pause), "--context", mgmtClusterKubeCtx, "--kubeconfig", mgmtClusterKubeconfigPath),
+		exec.WithStdout(GinkgoWriter),
+	)
+	err := command.RunAndRedirectOutput(context.Background())
 	Expect(err).ToNot(HaveOccurred())
 }
 

